DM-29617: Experimentation with remote registry

timj opened the pull request.

timj committed:
Initial stub remote registry subclass

timj committed:
Temporary fix for http without webdav

A proper fix will be done on DM-29708.

timj committed:
Add httpx to dependencies

timj committed:
Experiment with explicit pydantic models for client/server communication

timj committed:
Change to_simple for DataCoordinate to include full dataId

If the full information is present it should be used.

timj committed:
Add support for queryDatasets to remote registry

timj commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -0,0 +1,114 @@
+        return cls(**params)
+
+
+class QueryDatasetsModel(BaseModel):
```
@TallJimbo this is what I've started doing for modeling the parameters to the registry APIs.

timj committed:
Fix typo in docstring

timj committed:
Change globToRegex to prefer returning simple strings.

For "calexp" now returns "calexp" but for "cal*" returns
a regex as before. This should speed up lookups for the
general case of a user not giving a glob pattern.

timj committed:
Change DimensionRecord and DataCoordinate simple form to pydantic

This is necessary to allow DataCoordinate to include the
full dimension record information.

timj committed:
Change Registry ABC for queryDatasets to return iterable

Normally we return a special class that can iterate over
records and expand them on the fly. For now, revert to
allowing a simple iterator.

timj committed:
Try to use pydantic models specific to each dimension record type

timj committed:
Add queryDimensionRecords remote registry

timj commented on python/lsst/daf/butler/core/dimensions/_records.py:
```diff
@@ -75,6 +78,61 @@ def _subclassDimensionRecord(definition: DimensionElement) -> Type[DimensionReco
+        if not isinstance(v, collections.abc.Mapping):
+            raise ValueError(f"record can either be SpecificDimensionRecord or dict, not {v!r}")
+
+        universe = DimensionUniverse._instances[values["universe"]]
```
@TallJimbo I managed to get pydantic to create specific subclasses of serialized records so that validation could be enabled. The main problem being that you need a Universe to get the relevant DimensionElement.  This works because I store the universe version in the model and can therefore get the universe from the singleton. This can potentially significantly simplify the general `from_simple` methods since we could embed the universe version in all the models and they would ask Universe for that version (I assume that other versions should always fail because those versions aren't the one stored in the registry table). Thoughts:

* Currently I'm assuming that we will pass the version number from the client to the server in a parameters model. The server will compare that version to its own version and then if they match would pass that universe to `from_simple` as now.
* Embedding universe version potentially leads to the universe version number appearing at every level of the serialization and in one client/server interaction it should be the same everywhere.
* If we aren't embedding version numbers how could the dimension record be associated with the right universe? Since DimensionUniverse is already a singleton could we have an API that declares the active universe? (maybe as a context manager). One issue is that the model is instantiated by fastapi before we see it so I think I might need to create a universe  when the server starts up since otherwise there might not be one active if the records are being sent from the client (which means a context manager wouldn't work either).
* Currently the version number of a universe is a private attribute. There needs to be a public API for it.
* How would we distinguish a SPHEREx universe version from a Rubin universe version?

TallJimbo replied:
I'm not actually sure the extra automatic validation is worth it in this case.  I figured having separate classes for different dimensions was useful (as opposed to `SimpleNamespace` or `dict`) because there typos are a real problem and tab-completion of attributes is nice, and I was worried (paranoid?) about using a class with `__slots__` instead of `__dict__` storage for memory efficiency.  But I'm not sure that's really won out over the complexities of having dynamically-defined types even in the pure local `SqlRegistry` case.

And while we may not be able to rely on all clients giving the server consistent dimensions and want to do some validation there, surely there's some pydantic hook that lets you e.g. validate that the keys of a dict match some expectation?  (And if not, I imagine we could always explicitly check in the server code).

I'm already pretty uncomfortable with using the singleton dict keyed on version to work around the pickle limitations in `mulitprocessing` (if that let me use a custom `Unpickler`, I absolutely would, and drop the singleton entirely), as I think it dangerously conflates the kind of version number we want for schema management with something that's really an ugly implementation detail.  It'd be safer to use some kind of digest/hash instead of an incrementing number, at least in the singleton, and maybe elsewhere, at least for everything other than linear, "dimensions used by the Rubin product" human-meaningful schema versioning (which already has problems with e.g. SPHEREx, as you note).

timj replied:
> surely there's some pydantic hook that lets you e.g. validate that the keys of a dict match some expectation

I think all validation requires a universe and the pydantic models don't have a universe (if they did everything would work anyhow). I think we'd have to put the validation directly in `from_simple` since that does have a universe. I will switch to that approach.

timj replied:
I retained the dynamic pydantic model creation and now create a specialist class to validate everything in `from_simple`. It works well and if I send odd data types or extra fields it does now complain. An obvious problem is that I dynamically create a model every time I call `from_simple` on a record and that seems less than ideal. Caching it works but the hash for DimensionElement is the name of the element so that doesn't seem safe because the universe can change -- should the DimensionElement know the universe it came from?

TallJimbo replied:
`DimensionElement.universe` does exist, so you could cache on that plus the name (if `DimensionUniverse` is hashable; I'm not sure about that).

I wouldn't add the unvierse to `DimensionElement.__hash__` because that's allowed to be equal for non-equal objects anyway; adding a universe comparison to `DimensionElement.__eq__` might be okay, but there's also all of the `Named*` containers that assume that name equality is equal to object equality (ooh, that might be a problem for non-unique DatasetType names...), so I'd prefer to just define the equality comparison as "only meaningful within a universe" and add the universe explicitly only in contexts (like that cache) where necessary.

timj replied:
Thanks. This seems to work (including the universe).

timj committed:
Use Field to document models

timj committed:
Add remote registry query-data-ids

timj committed:
Allow query models to handle None kwargs

timj committed:
Change CollectionSearch to a pydantic model

timj committed:
Support remote butler query-collections command line

TallJimbo commented on python/lsst/daf/butler/core/utils.py:
```diff
@@ -408,23 +408,40 @@ def findFileResources(values: Iterable[str], regex: Optional[str] = None) -> Lis
         return Ellipsis
-    return [re.compile(fnmatch.translate(e)) for e in expressions]
+
+    nomagic = re.compile("^[A-Za-z_]+$")
```
If this is ever applied to collection expressions we probably want to include `/` as well.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+        self._dimensions: Optional[DimensionUniverse] = None
+
+        # Does each API need to be sent the defaults so that the server
+        # can use specific defaults each time?
```
I think I'd try first to just make the defaults purely client-side and never rely on the defaults in the server.

timj replied:
This seems doable but the downside then is that the client has to reimplement the server logic of knowing when to call standardize or default the collection and run. If the client sends everything to the server this all happens without the client having to care. On the other hand doing it client side would simplify the interface and potentially reduce the size of the JSON. Although non-python clients won't have any idea how to call `DataCoordinate.standardize()`.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+        # Dimensions should be pulled from server.
+
+    def __str__(self) -> str:
+        return str(self._db)
```
Make sure this doesn't repeat [DM-29249](https://jira.lsstcorp.org/browse/DM-29249) (which I'm now reminded we still haven't fixed).

timj replied:
Auth is going to be done with tokens so I don't think this is going to be an issue here.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+        return str(self._db)
+
+    def __repr__(self) -> str:
+        return f"RemoteRegistry({self._db!r}, {self.dimensions!r})"
```
Probably want `writeable` here too.

timj replied:
SqlRegistry does not report writeable so I was following that.

TallJimbo replied:
:+1:  I'm not sure if that was intentional or not.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+            raise e
+        except httpx.HTTPStatusError as e:
+            print(f"Error response {e.response.status_code} while requesting {e.request.url!r}.")
+            raise e
```
Is this kind of `print` logging desirable in this context for some reason, or are these debug relics?

timj replied:
I cleaned that up over the weekend. It was the first client routine I wrote so it had some debug examples in there from the httpx documentation.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+        try:
+            yield
+        except BaseException:
+            raise
```
As we've discussed a bit, I don't think there's any way out of this without changing the `SqlRegistry` interface.

In the meantime, if you can make this `raise NotImplementedError` or something to make sure nobody thinks it'll do something it doesn't, that'd be better, I think. But I understand if that's already impossible.

timj replied:
I had it be a no-op so that it would let the client experiment with doing writes on the server without having to solve the hard problem.  Another way to look at it is that this might lead to us getting in to trouble later on when we do want to think about transaction handling.  I'll change it to raise for the moment.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+
+        # Need to determine what to refresh.
+        # Might need a server method to return all the DatasetTypes up front.
+        pass
```
It occurs to me that this is not really a safe thing to have in the API at all; even if there is no client state to refresh (yet), it should tell the server to refresh its in-memory state from the DB, so it can pick up e.g. new dataset type registrations from other server instances.  But we also really don't want each random client to be able to tell the server to do a full refresh.

In many places, we could probably do automatic refreshes when e.g. looking for something and we don't see it in the cache.  But we probably need to worry about making those incremental so we don't get O(N^2) scaling or similar from repeated cache misses and total refreshes.  And there are also places where we really can't know that the cache is incomplete, anyway, such as when iterating over dataset types or collections in order to generate a larger query.

I don't see a way out of this that isn't really difficult.  Might be best to start by adding a suite of `Manager` classes that never hold any kind of local state (that should be doable within the current APIs) to just avoid all of the caching complexity imposed by the current managers, and start over again at optimizing things based on what's actually slow (which will often put the caches in the client instead of the server `Registry` instance, though we should be careful not to optimize for the one-concurrent-user case).  Anyhow, that's definitely not a project for this ticket, but it's one of the big hard problems we'll have to solve.

timj replied:
pipetask uses `registry.refresh` in multiple places so I felt I had to include the method for now.  I am happy to remove it from the ABC but that would make it impossible for pipetask use remote registry in the medium term (maybe that's not going to be a problem given that we'd first have to sort out the datastore/remote registry interaction).

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+    # insertOpaqueData + fetchOpaqueData + deleteOpaqueData
+    #    There are no managers for opaque data in client. This implies
+    #    that the server would have to have specific implementations for
+    #    use by Datastore. DatastoreBridgeManager also is not needed.
```
I did have designs on being able to use the opaque tables for other things (e.g. backing a `SqlDatastore`'s actual dataset storage, not just its extra records, and using that to save SQuaSH metrics), but I'm not sure yet whether any of them are viable in the no-shared-DB execution environment.  So, yes, let's not add those methods here now, but I'd rather not rule out having something like them in the future.

timj replied:
I think I'm edging towards having remote datastore registry provide higher level interfaces but it may well be that this interface does eventually turn up directly on server.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+    #    use by Datastore. DatastoreBridgeManager also is not needed.
+
+    # Should makeQueryBuilder be a public method? Nothing uses it outside
+    # of Registry.
```
Does not need to be public.  If I can ever get back to working on the query system it would be wholesale replaced anyway.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -0,0 +1,431 @@
+        """
+        from ..tests._dummyRegistry import DummyDatastoreRegistryBridgeManager, DummyOpaqueTableStorageManager
+        return DummyDatastoreRegistryBridgeManager(DummyOpaqueTableStorageManager(),
+                                                   self.dimensions, int)
```
I suspect I'll know more when I've looked at more of the PR, but I'm guessing it might make more sense to just move this method to `SqlRegistry` and not define it for `RemoteRegistry`.

timj replied:
Yes, I added it here for now because without it *everything* blows up because Butler tries to create a Datastore class by passing in the manager and this requires a "real" datastore manager even if Datastore is not being used. Without this I couldn't even demonstrate the `butler query-X` command lines worked.

TallJimbo commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -0,0 +1,107 @@
+
+# DataId and kwargs can only take limited types
+DataIdValues = Union[str, int, float]
+DataId = Mapping[str, DataIdValues]
```
There are already some very similar (but not quite identical) type aliases at

https://github.com/lsst/daf_butler/blob/f94a57823ca904996b74b8bab85282891f2d18a5/python/lsst/daf/butler/core/dimensions/_coordinate.py#L62

and

https://github.com/lsst/daf_butler/blob/f94a57823ca904996b74b8bab85282891f2d18a5/python/lsst/daf/butler/core/dimensions/_coordinate.py#L685

timj replied:
Yes, I saw those and I think I changed this code over the weekend. I thought I had pushed before I sent out the review request but clearly I hadn't. I did change it to drop the `float` but should have added the `None`.

It's funny that both those definitions for DataId value are in the same file.

TallJimbo commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -0,0 +1,107 @@
+    """List of regular expression strings."""
+
+    glob: Optional[List[str]] = None
+    """List of dataset type names or globs."""
```
I'm a bit surprised to see globs mentioned at all here; I had considered those to be purely part of the CLI.

timj replied:
I'm starting to get concerned that you are reviewing an old version of this pull request somehow.

Anyhow, yes, I called it `glob` rather than `strings` because when I was using the test server directly (and not as part of the RemoteRegistry) it was easier to be able to say `cal*` to do the wildcard matching. There will be some users of the server who aren't using Python.

TallJimbo replied:
Thinking about this a bit more, maybe we make `CollectionSearch.fromExpression` (and similar) responsible for identifying glob-like strings and turning them into regexes.  It's already responsible for standardizing a bunch of other kinds of inputs, and that'd let us use the globs in the Python API all over the place.

TallJimbo commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -98,10 +104,11 @@ class QueryDatasetsModel(BaseModel):
     bind: Optional[DataId] = None
     check: bool = True
-    kwargs: Optional[DataId] = None
+    keyword_args: Optional[DataId] = None  # mypy refuses to allow kwargs in model
```
This problem goes away if you do a little bit more standardization/normalization of the arguments on the client before or when making the model, such as calling `DataCoordinate.standardize` to merge `dataId` and `**kwargs` to generate something that's definitely a `DataCoordinate`.  Could go in a `classmethod` factory on the model class.

timj replied:
I had forgotten that I can call `DataCoordinate.standardize` without a registry and so can fold those into the DataId before I send it.

timj replied:
In almost every case sql registry calls `self.expandDataId` on the arguments. I think a `Registry.standardizeDataId()` method that folds in dataId+kwargs+defaults would work. It doesn't look like RegistryDefaults is amenable to pydantic because it (1) takes `**kwargs` and (2) needs registry to finish construction (making it incompatible with being immutable). I'd need to make a to_simple/from_simple and also allow `RegistryDefaults` to be constructed from a trusted dataId+collections+run rather than requiring a registry to check the dataId.

TallJimbo commented on python/lsst/daf/butler/registry/_remoteRegistry.py:
```diff
@@ -382,7 +386,36 @@ def queryDatasets(self, datasetType: Any, *,
         # Docstring inherited from lsst.daf.butler.registry.Registry
-        raise NotImplementedError()
+        if collections is not None:
+            print("Collections:", collections)
```
Debug relic.

timj replied:
That was removed yesterday.

TallJimbo commented on python/lsst/daf/butler/core/dimensions/_coordinate.py:
```diff
@@ -59,12 +60,20 @@
 
-DataIdValue = Union[str, int, None]
+# Pydantic will cast int to str if str is first in the Union.
+DataIdValue = Union[int, str, None]
```
Another duplicate of https://github.com/lsst/daf_butler/blob/f94a57823ca904996b74b8bab85282891f2d18a5/python/lsst/daf/butler/core/dimensions/_coordinate.py#L62

timj replied:
It's the actual line but you have to put the int before the str to control Pydantic casting.

TallJimbo replied:
Feel free to change the one in `_coordinate.py` to make pydantic happy; I'm pretty sure it's the only thing that cares about the order right now.

timj replied:
We are commenting on the line in `_coordinate.py`. This change fixes the order.

TallJimbo replied:
:facepalm:

TallJimbo commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -82,33 +93,62 @@ def from_expression(cls, expression: Any) -> ExpressionQueryParameter:
+            elif isinstance(expression, str):
                 params["glob"].append(expression)
+            elif hasattr(expression, "name"):
+                params["glob"].append(expression.name)
```
Level of duplicative logic is starting to worry me here, but I think there's a commit I haven't gotten to yet starts to address that...

timj replied:
Oh, you are reviewing commit by commit. That is not a good idea on this pull request where much of the code is work in progress building up to something that works as I experiment and keep cleaning up as I learn more. I had not expected you to be reviewing like this.

timj replied:
This doesn't seem like duplication though. It supports someone passing in a DatasetType but me wanting to send the name of that dataset type over the wire (and not the full DatasetType serialized form)

TallJimbo replied:
> Oh, you are reviewing commit by commit. That is not a good idea on this pull request

Yeah, I realized that you didn't expect this last night while reading your replies, a couple of threads before you realized that's what I'd done.  In any case, aside from these false positives (I'm going through and resolving all of the clear-cut ones I see; feel free to resolve any others yourself), I think it may have actually helped me wrap my head around it a bit as it gave me a chronological view rather than an alphabetical view of a big suite of changes.

TallJimbo replied:
And yes, I just misunderstood what this was doing r.e. duplication - it just looked suspiciously like the code that standardizes dataset type expressions as query inputs, but of course it's doing something quite different.

timj replied:
although given your other ticket it sounds like I should be adding a `datasetTypes` field to the model and storing the full form for transmission to the server so it can work out exactly which dataset types it is going to be using and not just the name.

TallJimbo replied:
Good point.  FWIW, if the user just passed the name you wouldn't need to pass more than that to the server (because the server could do the expansion), but if they did pass more, it would become important to respect that.  And it might make sense to always do the expansion on the client anyway.

TallJimbo commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -82,33 +93,62 @@ def from_expression(cls, expression: Any) -> ExpressionQueryParameter:
+    collections: Optional[ExpressionQueryParameter] = None
+    where: Optional[str] = None
+    components: Optional[bool] = None
+    bind: Optional[SimpleDataId] = None
```
This is not a data ID; the mapping is from `str` to any type that a `DimensionRecord` attribute can have (so at least also `bool`, `float`, and `Timespan`, in addition to `int` and `str`).

TallJimbo reviewed (approved):
Looks good. Lots of comments, but nothing major that isn't a "think about for the future" thing; closest is that I'd like to avoid using the `DimensionUniverse` anywhere we have any other option, even if that means not having pydantic validation of `DimensionRecords`.

I'm about to disappear until later this evening, so while I haven't forgotten about the questions in Jira, it'll be a bit before I get to them.  I also haven't looked at the server code at all yet; I'll do that later, too.

timj committed:
Fix some types in server models

timj committed:
Reorganize DimensionRecord serialization validation

Still dynamically create a model based on the definition
but now only use it in from_simple where we already
have a universe to validate against.

timj committed:
Add a cache for dimension record pydantic class creation

timj committed:
Reorganize pluggable registry location

Since registries are now imported dynamically they should
not be imported by default. Move the two registry classes
to a new `registries` location.

timj commented:
@TallJimbo FYI I've relocated the SqlRegistry and RemoteRegistry to `lsst.daf.butler.registries` matching the pluggable datastore approach. These registries should not be imported by default but should instead be loaded via configuration.

timj committed:
Disable transaction creation in remote registry

We do not use the facility yet and if we do we need to think
more about it.

timj committed:
Make registry.makeQueryBuilder a private method

timj committed:
Add v1 to REST API

Now uses "butler/v1/registry"

rra commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -0,0 +1,286 @@
+    """Base model for all query models."""
+
+    @validator("keyword_args", check_fields=False)
+    def check_keyword_args(cls, v, values) -> Optional[SimpleDataId]:  # type: ignore # noqa: N805
```
Nit: I tend to name pydantic validators with a leading underscore since they're not intended to be part of the public interface of the class.

rra commented on python/lsst/daf/butler/core/serverModels.py:
```diff
@@ -0,0 +1,286 @@
+    components: Optional[bool] = Components
+    bind: Optional[BindType] = Bind
+    check: bool = Check
+    keyword_args: Optional[SimpleDataId] = KeywordArgs  # mypy refuses to allow kwargs in model
```
Can't you lift this up to the parent class since it's common in all subclasses? Or is that what the comment is about?

timj replied:
I was going to build all the models and then work out which bits needed to be in the base.

rra reviewed (commented):
Nothing jumps out at me as unreasonable here. I'm surprised you managed to get defining fields at the top level to work; I struggled with that for a while but mypy really didn't like it. I may steal your approach.

rra commented on python/lsst/daf/butler/registries/remote.py:
```diff
@@ -0,0 +1,522 @@
+        # This could use a local cache since collection types won't
+        # change.
+        path = f"v1/registry/collection/type/{name}"
+        response = httpx.get(str(self._db.join(path)))
```
Does `ButlerURI` enforce that the server URI ends in a `/`? If not, you may want to enforce that.

timj replied:
It's annoyingly complicated in general but yes, ButlerURI knows if it represents a "directory".

rra reviewed (commented):
Yeah, the general shape of `remote.py` looks right to me. That's what I'd expect of the translation error: convert parameters to a pydantic model and then make the call and convert the JSON back to a model.

timj committed:
Add CollectionType.from_name(s) API

We had a few places where code was looking inside __members__
and it's much cleaner to do that internal to the class.

timj committed:
Use a httpx.Client to allow reuse of connection

timj committed:
Change getCollectionChain interface in server

timj committed:
Fake a version for mypy action

Since version.py is not created here.

timj committed:
Hide the pydantic validators from public interface
